### PR TITLE
Nft error recovery

### PR DESCRIPTION
**Pull request by tbezman**
### Architecture

This image includes every place on Gallery that we display Nfts / Nft Previews. We now have an `NftFailureBoundary` which catches errors beneath them and displays an appropriate fallback.

image

### `useNftRetry`

This hook is responsible for the following.

1. Receiving the failure event from the error boundary
2. Receiving the load event from the corresponding image / video (this may be simplified with suspense)
3. Refreshing the metadata when appropriate
4. Tracking whether or not we're in a failure state
5. Providing a `retryKey` to remount a component when necessary
6. Reporting errors to Sentry

### Other Changes

- `graphqlGetResizedNftImageUrlWithFallback` and `getVideoOrImageUrlForNftPreview` no longer return fallback URLs. If we could not get an image, we'll put the image into a failure state.
- `useSetContentIsLoaded` is now fully controlled by `useNftRetry`'s `handleNftLoad` callback. This way we don't have to have these things littered around everywhere. **Anything that loads an NFT should have a backing `useNftRetry` so we should always have full coverage.**

#### Gallery Page
image

#### Organize Gallery
image

#### Collection Editor
image

#### NftDetail
image

### Testing

Run this SQL against a database of your choice
```sql
UPDATE
	tokens
SET
	media = jsonb_set(media, '{media_url}', '"https://somescuffedimage.com"')
WHERE
	id = 'YOUR_TOKEN_ID_HERE';
```

### CODE REVIEW

File: src/components/LoadingAsset/ImageWithLoading.tsx [hunk 1]
```diff
@@ -16,6 +16,7 @@ type Props = {
   alt: string;
   heightType?: ContentHeightType;
   onClick?: () => void;
+  onLoad: JSX.IntrinsicElements['img']['onLoad'];
```
Comment (Robinnnnn) [inline src/components/LoadingAsset/ImageWithLoading.tsx:19]: 😍

File: src/components/NftFailureFallback/NftFailureBoundary.tsx [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+type State = {
+  error: Error | null;
+};
+
+type Props = PropsWithChildren<{
+  fallback: ReactNode;
+  onError: (error: Error) => void;
+}>;
+
+export class NftFailureBoundary extends Component<Props, State> {
```
Comment (Robinnnnn) [inline src/components/NftFailureFallback/NftFailureBoundary.tsx:12]: this is sickkkkk

File: src/components/NftFailureFallback/NftFailureFallback.tsx [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+      return;
+    }
+
+    onRetry?.();
+  }, [noControls, onRetry, refreshing]);
+
+  const spaceY = {
+    tiny: 4,
+    medium: 16,
+  }[size];
```
Comment (Robinnnnn) [inline src/components/NftFailureFallback/NftFailureFallback.tsx:33]: 2 cool 4 school

File: src/components/NftPreview/NftPreview.tsx [hunk 4]
```diff
@@ -78,8 +89,18 @@ function NftPreview({
     tokenRef
   );
 
+  const { collection, tokenSettings } = collectionToken;
+
   const username = collection.gallery?.owner?.username;
   const collectionId = collection.dbid;
+
+  // This fragment is split up, so we can refresh it as a part
+  // of the retry system
```
Comment (Robinnnnn) [inline src/components/NftPreview/NftPreview.tsx:98]: 🤯

File: src/errors/CouldNotRenderNftError.ts [hunk 5]
```diff
@@ -0,0 +1,13 @@
+import { Primitive } from 'relay-runtime/lib/store/RelayStoreTypes';
+
+export class CouldNotRenderNftError extends Error {
```
Comment (Robinnnnn) [inline src/errors/CouldNotRenderNftError.ts:3]: HOT

File: src/hooks/useNftRetry.tsx [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+type useNftRetryResult = {
+  retryKey: number;
+  isFailed: boolean;
+  refreshMetadata: () => void;
+  refreshingMetadata: boolean;
+  handleNftLoaded: ContentIsLoadedEvent;
+  handleNftError: (error: Error) => void;
+};
+
+export function useNftRetry({ tokenId }: useNftRetryArgs): useNftRetryResult {
```
Comment (Robinnnnn) [inline src/hooks/useNftRetry.tsx:24]: this is ridiculous

File: src/hooks/useNftRetry.tsx [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+
+      if (error instanceof CouldNotRenderNftError) {
+        reportError(error.message, {
+          tags: { ...commonTags, ...error.metadata },
+        });
+      } else {
+        reportError('Could not load nft', {
+          tags: commonTags,
+        });
+      }
```
Comment (Robinnnnn) [inline src/hooks/useNftRetry.tsx:72]: TODO: come up with a process to routinely check on these errors

**Review by Robinnnnn — APPROVED** [review]
🔥🔥🔥🔥🔥🔥

File: src/hooks/useNftRetry.tsx (hunk 7 above)
Comment (tbezman) [reply src/hooks/useNftRetry.tsx]: I added a prefix to report error so they should be easy to create a search for

**Pull request merged**